include/deal.II/lac/constrained_linear_operator.h

```diff
@@ -0,0 +1,292 @@
+#define dealii__constrained_linear_operator_h
+
+#include <deal.II/lac/linear_operator.h>
+#include <deal.II/lac/packaged_operation.h>
+#include <deal.II/lac/constraint_matrix.h>
+
+#ifdef DEAL_II_WITH_CXX11
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * @name Indirectly applying constraints to LinearOperator
+ */
+//@{
+
+
+/**
+ * This function takes a ConstraintMatrix @p constraint_matrix and an
+ * operator exemplar @p exemplar (this exemplar is usually a linear
+ * operator that describes the system matrix) and returns a LinearOperator
+ * object associated with the "homogeneous action" of the underlying
+ * ConstraintMatrix object:
+ *
+ * Applying the LinearOperator object on a vector <code>u</code> results in
+ * a vector <code>v</code> that stores the result of calling
+ * ConstraintMatrix::distribute() on <code>u</code> - with one important
+ * difference: inhomogeneities are note applied, but always treated as 0
+ * instead.
+ *
+ * The LinearOperator object created by this function is primarily used
+ * internally in constrained_linear_operator() to build up a modified
+ * system of linear equations. How to solve a linear system of equations
+ * with this approach is explained in detail in the @ref constraints
+ * module.
+ *
+ * @author Mauro Bardelloni, Matthias Maier, 2015
+ *
+ * @relates LinearOperator
+ * @ingroup constraints
+ */
+template <typename Range, typename Domain>
+LinearOperator<Range, Domain> distribute_constraints_linear_operator(
+  const ConstraintMatrix &constraint_matrix,
+  const LinearOperator<Range, Domain> &exemplar)
+{
+  LinearOperator<Range, Domain> return_op = exemplar;
+
+  return_op.vmult_add = [&constraint_matrix](Range &v, const Domain &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      {
+        if (constraint_matrix.is_constrained(i))
+          {
+            const auto *entries = constraint_matrix.get_constraint_entries (i);
+            for (types::global_dof_index j = 0; j < entries->size(); ++j)
+              {
+                const auto pos = (*entries)[j].first;
+                v(i) +=  u(pos) * (*entries)[j].second;
+              }
+          }
+        else
+          v(i) += u(i);
+      }
+  };
+
+  return_op.Tvmult_add = [&constraint_matrix](Domain &v, const Range &u)
+  {
+    for (auto i : v.locally_owned_elements())
+      {
+        if (constraint_matrix.is_constrained(i))
+          {
+            const auto *entries = constraint_matrix.get_constraint_entries(i);
+            for (types::global_dof_index j = 0; j < entries->size(); ++j)
+              {
+                const auto pos = (*entries)[j].first;
+                v(pos) += u(i) * (*entries)[j].second;
+              }
+          }
+        else
+          v(i)+=u(i);
+      }
+  };
+
+  // lambda capture expressions are a C++14 feature...
+  const auto vmult_add = return_op.vmult_add;
+  return_op.vmult = [vmult_add](Range &v, const Domain &u)
+  {
+    v = 0.;
+    vmult_add(v, u);
```

Comment:
This too is not wrong, but slow -- for every write operation you have to read v once. It shouldn't be too hard to specialize the algorithm above for a zero starting vector.